[Input] Stop accumulation of duplicate InputActions (case ISX-1902)

AlexTyrer opened the pull request:
### Description

FIX: Stop the accumulation of duplicate Actions being created when loading/unloading the Input System test scenes.

### Changes made

o ProjectWideInputActions are now only enabled in playmode - the state is controlled by EditorApplication.playModeStateChanged listener PlayStateNotifier.ModeChanged()

o Only create DefaultInputActions once in AssignDefaultActions() (controlled with class static defaultActions)

o Improve Input Debug window Action item display (show which action map asset each action was created from)


### Notes

InputSystem.actions are now only enabled in playmode.


### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

AlexTyrer committed:
[Input] Stop accumulation of duplicate InputActions (case ISX-1902)

o ProjectWideInputActions are now only enabled in playmode ...
  ... state is controlled by EditorApplication.playModeStateChanged listener PlayStateNotifier.ModeChanged()
o Only create DefaultInputActions once in AssignDefaultActions() (controlled with class static defaultActions)
o Improve Input Debug window Action item display (show which action map asset each action was created from)

AlexTyrer committed:
[Input] Fix for !editor builds / fix code formatting

o Wrap ModeChanged() listener class in #if UNITY_EDITOR
o Formatting: run Unity format tools on InputSystem.cs

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/Debugger/InputDebuggerWindow.cs:
```diff
@@ -927,6 +927,15 @@ private unsafe void AddActionItem(TreeViewItem parent, InputAction action, ref i
                 var name = action.actionMap != null ? $"{action.actionMap.name}/{action.name}" : action.name;
                 if (!action.enabled)
                     name += " (Disabled)";
+                if (action.actionMap != null && action.actionMap.m_Asset != null)
```
I think this is a helpful change. Since its also user facing I think it should be added under "Added" title in CHANGELOG.

AlexTyrer replied:
Will add those lines.

ekcoh replied:
thanks

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/InputSystem.cs:
```diff
@@ -3011,6 +3011,41 @@ internal static bool ShouldDrawWarningIconForBinding(string bindingPath)
 
+#if UNITY_EDITOR
+        [InitializeOnLoad]
+        public static class PlayStateNotifier
```
Its likely fine to keep this isolated but I also noticed we have both monitoring and evaluations of playmode state changes in https://github.com/Unity-Technologies/InputSystem/blob/develop/Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs#L341-L362 and https://github.com/Unity-Technologies/InputSystem/blob/develop/Packages/com.unity.inputsystem/InputSystem/InputManager.cs#L323

Wanted to share just for awareness. Might be better to keep this isolated.

AlexTyrer replied:
I see that other option now, thanks - was just focussed on InputSystem in isolation.

ekcoh replied:
I think we should keep it isolated, will resolve this

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/InputSystem.cs:
```diff
@@ -3011,6 +3011,41 @@ internal static bool ShouldDrawWarningIconForBinding(string bindingPath)
+                switch (playModeState)
+                {
+                    case PlayModeStateChange.EnteredEditMode:
+                        actions?.Disable();
```
Was there a reason for using this rather than ExitingPlayMode?

ekcoh replied:
Just wondering if there might be any implications of doing it after the mode change "edge"?

AlexTyrer replied:
I tried both ways and didn't see any difference.

It wasn't immediately clear to me if one was preferable to the other.

From reading the docs perhaps I should make this change to: EnteredPlayMode and ExitingPlaymode.

I can see an argument for both ways - it's unclear which are the most "balanced" - the enums and the comments about the enums balance differently.

AlexTyrer replied:
The **Entered*** calls are the right ones I think - see my comments below.

ekcoh replied:
This should ideally be covered in tests and I will try to cover that as part of modifying the tests which I currently do on a pending PR. Will resolve this.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/InputSystem.cs:
```diff
@@ -3011,6 +3011,41 @@ internal static bool ShouldDrawWarningIconForBinding(string bindingPath)
+                EditorApplication.playModeStateChanged += ModeChanged;
+            }
+
+            static void ModeChanged(PlayModeStateChange playModeState)
```
If not already checked it would be good to investigate and document along with the already added inline docs (which is great by the way) when the actions may be expected to be enabled. E.g.

Are they enabled when entering Awake() on a MonoBehavior?
Are they enabled when entering Start() on a MonoBehavior?
Are they enabled when entering OnEnable() on a MonoBehavior?
Are they disabled when entering OnDisable() on a MonoBehavior?
Are they disabled when entering OnDestroy() on a MonoBehavior? (When destroy by exiting play mode)

AlexTyrer replied:
I have added an explanatory to the source (see comment I made to @lyndon-unity [below](https://github.com/Unity-Technologies/InputSystem/pull/1863#discussion_r1508913457))

ekcoh replied:
Awesome, thanks, that is helpful

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModule.cs:
```diff
@@ -1388,6 +1394,7 @@ public void AssignDefaultActions()
         /// <seealso cref="AssignDefaultActions"/>
         public void UnassignActions()
         {
+            defaultActions = default;
```
If assigning null to default actions like this it seems like we will leak memory. DefaultInputActions seems to implement IDisposable so should be?
```

defaultActions?.Dispose();
defaultActions = default;
```

AlexTyrer replied:
Right - that's a good call, will change.

ekcoh reviewed (approved):
In general changes look good. I have some questions and some suggestions related to making sure its solid.

ekcoh commented:
It might be good for @lyndon-unity to review this PR since he was the may author of adding those enable() and disable() calls into InputSystem.cs in a previous PR.

ekcoh reviewed (changes requested):
Filing a request change for updating CHANGELOG.md, there should be one ADD entry for the Input Debugger addition, and a FIX entry for the accumulated actions from InputUIModule I believe.

AlexTyrer committed:
Merge branch 'develop' into ISX-1902/fix-input-action-accumulation

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/InputSystem.cs:
```diff
@@ -3011,6 +3011,41 @@ internal static bool ShouldDrawWarningIconForBinding(string bindingPath)
+                        break;
+
+                    case PlayModeStateChange.EnteredPlayMode:
+                        actions?.Enable();
```
I think MonoBehaviour's Start() is called before PlayModeStateChange.EnteredPlayMode
So when I (previously - not with this PR) tested to see if actions were enabled in my Start function, it was not enabled.
This is why I added the enable in 'InitializeInEditor' too. So it should already be enabled by here.
(history is in this PR : https://github.com/Unity-Technologies/InputSystem/pull/1834/files ) 

And another comment:
PlayModeStateChange.ExitingPlayMode gets hit before OnDisable 
(But I see you are using EnteredEditMode which is later)

lyndon-unity replied:
So I think there is a test in here to update:
Assets/Samples/ProjectWideActions/ProjectWideActionsExample.cs

lyndon-unity replied:
^ That test probably doesn't need updating as the Enable in the InitializeInEditor code is still present.

AlexTyrer replied:
OK, I've tested the various combinations here:

Using PlayModeStateChange.**EnteredEditMode** / PlayModeStateChange.**EnteredPlayMode**

```
    InputSystem.actions are
        Awake       enabled
        OnEnable    enabled
        Start       enabled
        OnDisable   enabled
        OnDestroy   enabled
```

Using PlayModeStateChange.**ExitingPlayMode** / PlayModeStateChange.**ExitingEditMode**

```
    InputSystem.actions are
        Awake       DISABLED
        OnEnable    DISABLED
        Start       DISABLED
        OnDisable   enabled
        OnDestroy   enabled
```
(they show as enabled in the OnDisable / OnDestroy as they are explicitly enabled in the sample's Start method if they are not already)
   
So I believe using the **Entered*** calls is the right thing to do here.

lyndon-unity reviewed (approved):
I think this looks good overall

ekcoh commented:
Thought about this a bit and generally this is just connecting an assets enable() and disable() logic to edit mode and play mode callbacks. I think we should at least consider (may be done after this PR lands) moving this out of InputSystem.cs and into InputActionAsset instead. This way, when the callbacks trigger we could easily just check if `InputSystem.actions == this` to determine if we should act. But more interestingly, it would open up for making this a setting/property on the asset instead and allow basically any InputActionAsset to have this behavior, while at the same time make it easy for users to opt-in or opt-out of that behavior both for project-wide or Any InputActionAsset.

ekcoh committed:
Merge branch 'develop' into ISX-1902/fix-input-action-accumulation

AlexTyrer committed:
Merge branch 'develop' into ISX-1902/fix-input-action-accumulation

AlexTyrer committed:
[Input] Address PR feedback / fix non-editor compilation / updated changelog

o Ensure that we call Dispose() on the default actions when in UnassignActions().
o Gated EditorApplication.isPlaying checks with #if UNITY_EDITOR
o CHANGELOG.md: Fixed (ISX-1902) / Added (InputDebug window Action improvement)

AlexTyrer commented:
> Filing a request change for updating CHANGELOG.md, there should be one ADD entry for the Input Debugger addition, and a FIX entry for the accumulated actions from InputUIModule I believe.

I've adde those here in commit **[b54385a](https://github.com/Unity-Technologies/InputSystem/pull/1863/commits/b54385a1692b0ccb92560921b662b5be0bf835e2)**

ekcoh commented on Packages/com.unity.inputsystem/CHANGELOG.md:
```diff
@@ -57,6 +58,7 @@ however, it has to be formatted properly to pass verification tests.
 - Fixed Input Actions Editor losing reference to current ControlScheme upon entering Play Mode [ISXB-770](https://issuetracker.unity3d.com/product/unity/issues/guid/ISXB-770).
 - Fixed Save shortcut (ctrl/cmd + S by default) not saving changes in Input Actions Editor windows. [ISXB-659](https://issuetracker.unity3d.com/product/unity/issues/guid/ISXB-659).
 - Fixed headers in InputActionEditor windows becoming squashed when there is a large number of Action Maps/Actions.
+- Fixed duplication of project wide input actions when loading/unloading scenes [ISX-1902]
```
Remove ISX-1902 from this since its an internal ticket. We only link public bug tickets ISXB and for them use issue tracker links.

ekcoh replied:
@AlexTyrer This one needs to be fixed before merge, but approved to not block this PR from landing.

AlexTyrer replied:
OK - I saw there are other some internal ISX- issues in the file (with no links) so assumed that was the thing to do.

Removed in 1da1021

AlexTyrer committed:
[Input] Fix formatting (arguable, but this is what the script does)

ekcoh commented:
@AlexTyrer this branch should also get some QA assessment I believe. Adding @Pauliusd01

AlexTyrer committed:
[Input] PR feedback - remove internal ticket number.

AlexTyrer commented:
> @AlexTyrer this branch should also get some QA assessment I believe. Adding @Pauliusd01

Good idea, thanks.

@Pauliusd01 you should just be able to run the samples and go in and out of all the scenes - you will see the actions in the "Input Debug" window come and go as each scene loads / unloads.

Exiting playmode should go back to the original state.

Note now that the project wide input actions are now **never** enabled in edit mode (only in playmode / the player) - so they won't be shown in the Input Debug window in edit mode.

Pauliusd01 commented:
This also fixes ISX-1602 which was closed for being low priority
